src/controller/java/AndroidKeyValueStoreManagerImpl.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <platform/KeyValueStoreManager.h>
+
+#include <algorithm>
+#include <string.h>
+
+#include "CHIPJNIError.h"
+#include "JniReferences.h"
+#include "JniTypeWrappers.h"
+
+#include <support/Base64.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace DeviceLayer {
+namespace PersistedStorage {
+namespace {
+
+constexpr size_t kMaxKvsValueBytes        = 1024;
+constexpr size_t kMaxKvsValueEncodedChars = BASE64_ENCODED_LEN(kMaxKvsValueBytes) + 1;
+
+} // namespace
+
+KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
+
+CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
+                                          size_t offset)
+{
+    ReturnErrorCodeIf(mEnv == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorCodeIf(mKeyValueStoreManagerClass == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorCodeIf(mGetMethod == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorCodeIf(offset != 0, CHIP_ERROR_INVALID_ARGUMENT);
+
+    UtfString javaKey(mEnv, key);
+
+    jobject javaValue = mEnv->CallStaticObjectMethod(mKeyValueStoreManagerClass, mGetMethod, javaKey.jniValue());
+    if (mEnv->ExceptionCheck())
+    {
+        ChipLogError(DeviceLayer, "Java exception in KVS::Get");
+        mEnv->ExceptionDescribe();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
+    }
+
+    if (javaValue == nullptr)
+    {
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+
+    JniUtfString utfValue(mEnv, (jstring) javaValue);
+    if (strlen(utfValue.c_str()) > kMaxKvsValueEncodedChars)
+    {
+        ChipLogError(DeviceLayer, "Unexpected large value received from KVS");
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    uint8_t buffer[kMaxKvsValueBytes];
+    uint16_t decodedLength = chip::Base64Decode(utfValue.c_str(), strlen(utfValue.c_str()), buffer);
```

Comment:
This would be a small optimization:

```suggestion
    uint16_t decodedLength = chip::Base64Decode(utfValue.c_str(), utfValue.length(), buffer);
```

Reply:
This is not critical.  No need to change.